File: apic.py
import taichi as ti
import taichi.math as tm
import numpy as np

ti.init(arch=ti.gpu, debug=False)

res_level = 1
n_particles = 8192 * 1
n_grid = tm.vec2(48 * res_level * 5, 48 * res_level)
dx = 1 / n_grid.y
domain_width = n_grid.x * dx
domain_height = n_grid.y * dx
dt = 2e-4 / res_level

p_rho = 1
p_vol = (dx * 0.5) ** 2
p_mass = p_vol * p_rho
gravity = 9.8
bound = 3
river_depth = 0.2
E = 400

x = ti.Vector.field(2, float, n_particles)
x_normalized = ti.Vector.field(2, float, n_particles)
v = ti.Vector.field(2, float, n_particles)
C = ti.Matrix.field(2, 2, float, n_particles)
J = ti.field(float, n_particles)

grid_v = ti.Vector.field(2, float, (n_grid.x, n_grid.y))
grid_m = ti.field(float, (n_grid.x, n_grid.y))

gui = ti.GUI("Eisbach", res=(3000, 300))
control_gui = ti.GUI("Control Panel", res=(200, 200))
inflow_rate_slider = control_gui.slider("inflow_rate", 1.0, 15.0, step=0.1)
inflow_rate_slider.value = 3.5

num_points = int(domain_width * 12)
riverbed_nodes_x = np.linspace(0, domain_width, num_points + 1, endpoint=True)
riverbed_nodes_y = ti.field(float, num_points + 1)
riverbed_nodes_y.from_numpy(np.array(
    [0.50, 0.50, 0.50, 0.50, 0.49, 0.49, 0.49, 0.49, 0.49, 0.49, 0.47, 0.41, 0.37, 0.35, 0.32, 0.30, 0.29, 0.29, 0.29,
     0.30, 0.31, 0.33, 0.36, 0.38, 0.48, 0.41, 0.37, 0.36, 0.34, 0.32, 0.32, 0.31, 0.45, 0.44, 0.44, 0.44, 0.44, 0.44,
     0.44, 0.44, 0.43, 0.43, 0.43, 0.43, 0.43, 0.43, 0.42, 0.42, 0.42, 0.42, 0.42, 0.41, 0.41, 0.41, 0.41, 0.41, 0.41,
     0.40, 0.40, 0.40, 0.40, ], dtype=np.float32))

riverbed_nodes = np.stack((riverbed_nodes_x, riverbed_nodes_y.to_numpy()), axis=1)
riverbed_nodes[:, 0] /= domain_width
riverbed_nodes[:, 1] /= domain_height

@ti.func
def riverbed(x, riverbed_nodes_y):
    """Returns the y-coordinate of the riverbed at position x, and the normal vector."""
    i = int(x * num_points / domain_width)  # [0, num_points-1]
    start_node_y = riverbed_nodes_y[i]
    start_node_x = i * (domain_width / num_points)
    end_node_y = riverbed_nodes_y[i + 1]
    end_node_x = (i + 1) * (domain_width / num_points)

    tangent = ti.Vector([end_node_x - start_node_x, end_node_y - start_node_y])
    normal = ti.Vector([-tangent.y, tangent.x]).normalized()

    if ti.abs(x - start_node_x) < dx:
        prev_node_y = riverbed_nodes_y[i - 1]
        prev_node_x = (i - 1) * (domain_width / num_points)
        prev_tangent = ti.Vector([start_node_x - prev_node_x, start_node_y - prev_node_y])
        normal += ti.Vector([-prev_tangent.y, prev_tangent.x]).normalized()
        normal = normal.normalized()
    elif ti.abs(x - end_node_x) < dx:
        next_node_y = riverbed_nodes_y[i + 2]
        next_node_x = (i + 2) * (domain_width / num_points)
        next_tangent = ti.Vector([next_node_x - end_node_x, next_node_y - end_node_y])
        normal += ti.Vector([-next_tangent.y, next_tangent.x]).normalized()
        normal = normal.normalized()

    y = start_node_y + tangent.y / tangent.x * (x - start_node_x)

    return y, normal


@ti.kernel
def substep(inflow: float, riverbed_nodes_y: ti.template()):
    for i, j in grid_m:
        grid_v[i, j] = [0, 0]
        grid_m[i, j] = 0
    for p in x:
        Xp = x[p] / dx
        base = int(Xp - 0.5)
        fx = Xp - base
        w = [0.5 * (1.5 - fx) ** 2, 0.75 - (fx - 1) ** 2, 0.5 * (fx - 0.5) ** 2]
        stress = -dt * 4 * E * p_vol * (J[p] - 1) / dx ** 2
        affine = ti.Matrix([[stress, 0], [0, stress]]) + p_mass * C[p]
        for i, j in ti.static(ti.ndrange(3, 3)):
            offset = ti.Vector([i, j])
            dpos = (offset - fx) * dx
            weight = w[i].x * w[j].y
            grid_v[base + offset] += weight * (p_mass * v[p] + affine @ dpos)
            grid_m[base + offset] += weight * p_mass
    for i, j in grid_m:
        if grid_m[i, j] > 0:
            grid_v[i, j] /= grid_m[i, j]
        grid_v[i, j].y -= dt * gravity

        # Inflow
        if i < bound and grid_v[i, j].x < 0:
            grid_v[i, j].x = inflow
        # outflow
        if i > n_grid.x - bound and grid_v[i, j].x > 0:
            grid_v[i, j].x = inflow
        # riverbed boundary
        xi = i * dx
        y_bound, normal = riverbed(xi, riverbed_nodes_y)
        y_j = int(y_bound * n_grid.y - 0.5) + 1
        normal_component = tm.dot(grid_v[i, j], normal)
        if j <= y_j and normal_component < 0:
            grid_v[i, j] -= normal_component * normal
            grid_v[i, j] *= 0.99

        if j > n_grid.y - bound and grid_v[i, j].y > 0:
            grid_v[i, j].y = 0
    for p in x:
        Xp = x[p] / dx
        base = int(Xp - 0.5)
        fx = Xp - base
        w = [0.5 * (1.5 - fx) ** 2, 0.75 - (fx - 1) ** 2, 0.5 * (fx - 0.5) ** 2]
        new_v = ti.Vector.zero(float, 2)
        new_C = ti.Matrix.zero(float, 2, 2)
        for i, j in ti.static(ti.ndrange(3, 3)):
            offset = ti.Vector([i, j])
            dpos = (offset - fx) * dx
            weight = w[i].x * w[j].y
            g_v = grid_v[base + offset]
            new_v += weight * g_v
            new_C += 4 * weight * g_v.outer_product(dpos) / dx ** 2
        v[p] = new_v
        x[p] += dt * v[p]
        J[p] *= 1 + dt * new_C.trace()
        C[p] = new_C

        # if the particle goes out of bounds, respawn it at the left side
        y, normal = riverbed(x[p].x, riverbed_nodes_y)
        if x[p].x > domain_width - 3 * dx or x[p].x < dx or x[p].y < y or x[p].y > domain_height - 3 * dx:
            x[p] = [ti.random() * 3 * dx + dx, ti.random() * river_depth]
            y, normal = riverbed(x[p].x, riverbed_nodes_y)
            x[p] += [0.0, y]
            v[p] = [normal.y, -normal.x]
            v[p] *= inflow
            J[p] = 1
            C[p] = ti.Matrix.zero(float, 2, 2)


@ti.kernel
def init(riverbed_nodes_y: ti.template()):
    for p in range(n_particles):
        x[p] = [ti.random() * 0.95 * domain_width + 3 * dx, ti.random() * river_depth]
        y, normal = riverbed(x[p].x, riverbed_nodes_y)
        x[p] += [0.0, y]
        v[p] = [normal.y * 2, -normal.x]
        J[p] = 1
        C[p] = ti.Matrix.zero(float, 2, 2)


init(riverbed_nodes_y)
start = control_gui.button('start')
stop = control_gui.button('stop')
save = control_gui.button('save')
score = control_gui.label('score')
running = False
while gui.running and not gui.get_event(gui.ESCAPE):
    if running:
        for s in range(200 * res_level):
            substep(inflow_rate_slider.value, riverbed_nodes_y)

    # Render riverbed nodes
    gui.clear(0x112F41)
    gui.circles(riverbed_nodes, radius=5.0, color=0xED553B)  # Draw riverbed
    gui.lines(riverbed_nodes[:-1], riverbed_nodes[1:], radius=1.0, color=0xED553B)

    particle_vis = x.to_numpy()
    particle_vis[:, 0] /= domain_width
    particle_vis[:, 1] /= domain_height
    gui.circles(particle_vis, radius=2.5, color=0x068587)

    if gui.is_pressed(ti.GUI.LMB):
        mouse_x, mouse_y = gui.get_cursor_pos()
        idx = int(mouse_x * num_points + 0.5)
        riverbed_nodes_y[idx] = mouse_y * domain_height
        riverbed_nodes[idx, 1] = riverbed_nodes_y[idx] / domain_height

    for e in control_gui.get_events():
        if e.key == save:
            print_str = ""
            for i in range(num_points + 1):
                print_str += f"{riverbed_nodes_y[i]:.2f}, "
            print(print_str)

        if e.key == start:
            running = True
        if e.key == stop:
            running = False
    gui.show()
    control_gui.show()

File: apic.js
let main = async () => {
  await ti.init();

  let quality = 1;
  //let n_particles = 9000 * quality ** 2;
  //let n_grid = 128 * quality;
  //let dx = 1 / n_grid;

    const res_level = 1;
    const n_particles = 8192 * 1;

// We use a simple object to mimic the vector structure
    const n_grid = { 
    x: 48 * res_level * 1 , 
    y: 48 * res_level 
    };

const dx = 1 / n_grid.y;
const domain_width = n_grid.x * dx;
const domain_height = n_grid.y * dx;
const dt = 2e-4 / res_level;

  let inv_dx = n_grid;
  let dt = 1e-4 / quality;
  let p_vol = (dx * 0.5) ** 2;
  let p_rho = 1;
  let p_mass = p_vol * p_rho;
  let E = 5e3; // Young's modulus a
  let nu = 0.2; // Poisson's ratio
  let mu_0 = E / (2 * (1 + nu));
  let lambda_0 = (E * nu) / ((1 + nu) * (1 - 2 * nu)); // Lame parameters
  let x = ti.Vector.field(2, ti.f32, [n_particles]); // position
  let v = ti.Vector.field(2, ti.f32, [n_particles]); // velocity
  let C = ti.Matrix.field(2, 2, ti.f32, [n_particles]); // affine vel field
  let F = ti.Matrix.field(2, 2, ti.f32, n_particles); // deformation gradient
  let material = ti.field(ti.i32, [n_particles]); // material id
  let Jp = ti.field(ti.f32, [n_particles]); // plastic deformation
  let grid_v = ti.Vector.field(2, ti.f32, [n_grid, n_grid]);
  let grid_m = ti.field(ti.f32, [n_grid, n_grid]);

  let n_nodes = 10;
  let ground_x_values = [...Array(n_nodes).keys()];
  let ground_y_values = ti.field(ti.f32, n_nodes);

  //from slider values:
  let x_val = 1;

  let para_a = 1;
  const para_a_slider_elm = document.getElementById("para_a");

  let para_b = 0.4;
  let kick_b = 0.2;
  let kick_h = 0.1;
  let kick_a = -10;

  let inflow = 10;
  let river_depth = 0.5;

  const bound = 3;

  const img_size = 512;
  let image = ti.Vector.field(4, ti.f32, [img_size, img_size]);
  const group_size = n_particles / 3;

  let riverbed = (x_val, para_a, para_b, kick_b, kick_h, kick_a) => {
    let y = 0.0;
    let normal = [0.0, 1.0];
    let parabola_c = 3 * dx;
    let ground_transition_x = 0.7;

    let kicker_c = para_a * (kick_b - para_b) ** 2 + 3 * dx + kick_h;
    let A = kick_a - para_a;
    let B = 2 * (para_a * para_b - kick_a * kick_b);
    let D = kick_a * kick_b ** 2 - para_a * para_b ** 2 + kicker_c - parabola_c;

    // Solve quadratic intersection
    let delta = B ** 2 - 4 * A * D;
    // Determine kicker range. Using dummy values if delta < 0 to avoid NaN
    let kicker_start = 0.0;
    let kicker_end = 0.0;

    if (delta >= 0) {
      kicker_start = (-B + ti.sqrt(delta)) / (2 * A);
      kicker_end = (-B - ti.sqrt(delta)) / (2 * A);
    }

    let ground_transition_y =
      para_a * (ground_transition_x - para_b) ** 2 + 3 * dx;

    if (x_val > kicker_start && x_val < kicker_end) {
      y = kick_a * (x_val - kick_b) ** 2 + kicker_c;
      let dy_dx = 2 * kick_a * (x_val - kick_b);
      let norm_raw = [-dy_dx, 1.0];
      normal = ti.normalized(norm_raw);
    } else if (x_val < ground_transition_x) {
      y = para_a * (x_val - para_b) ** 2 + parabola_c;
      let dy_dx = 2 * para_a * (x_val - para_b);
      let norm_raw = [-dy_dx, 1.0];
      normal = ti.normalized(norm_raw);
    } else {
      y = ground_transition_y;
      normal = [0.0, 1.0];
    }

    return { y, normal };
  };

  ti.addToKernelScope({
    n_particles,
    n_grid,
    dx,
    inv_dx,
    dt,
    p_vol,
    p_rho,
    p_mass,
    E,
    nu,
    mu_0,
    lambda_0,
    x,
    v,
    C,
    F,
    material,
    Jp,
    grid_v,
    grid_m,
    image,
    img_size,
    group_size,
    bound,
    riverbed,
    para_a,
    para_b,
    kick_a,
    kick_b,
    kick_h,
    inflow,
    river_depth,
  });

  let substep = ti.kernel({ f: ti.template() }, (para_a_slider, f) => {
    let test = f[0];
    for (let I of ti.ndrange(n_grid, n_grid)) {
      grid_v[I] = [0, 0];
      grid_m[I] = 0;
    }
    for (let p of ti.range(n_particles)) {
      let base = i32(x[p] * inv_dx - 0.5);
      let fx = x[p] * inv_dx - f32(base);
      let w = [
        0.5 * (1.5 - fx) ** 2,
        0.75 - (fx - 1) ** 2,
        0.5 * (fx - 0.5) ** 2,
      ];
      F[p] = (
        [
          [1.0, 0.0],
          [0.0, 1.0],
        ] +
        dt * C[p]
      ).matmul(F[p]);

      let h = f32(max(0.1, min(5, ti.exp(10 * (1.0 - Jp[p])))));
      if (material[p] == 1) {
        h = 0.3;
      }
      let mu = mu_0 * h;
      let la = lambda_0 * h;
      if (material[p] == 0) {
        mu = 0.0;
      }
      let svd = ti.svd2D(F[p]);
      let U = svd.U;
      let sig = svd.E;
      let V = svd.V;
      let J = 1.0;
      for (let d of ti.static(ti.range(2))) {
        let new_sig = sig[[d, d]];
        if (material[p] == 2) {
          // Plasticity
          new_sig = min(max(sig[[d, d]], 1 - 2.5e-2), 1 + 4.5e-3);
        }
        Jp[p] = (Jp[p] * sig[[d, d]]) / new_sig;
        sig[[d, d]] = new_sig;
        J = J * new_sig;
      }
      if (material[p] == 0) {
        F[p] =
          [
            [1.0, 0.0],
            [0.0, 1.0],
          ] * sqrt(J);
      } else if (material[p] == 2) {
        F[p] = U.matmul(sig).matmul(V.transpose());
      }
      let stress =
        (2 * mu * (F[p] - U.matmul(V.transpose()))).matmul(F[p].transpose()) +
        [
          [1.0, 0.0],
          [0.0, 1.0],
        ] *
          la *
          J *
          (J - 1);
      stress = -dt * p_vol * 4 * inv_dx * inv_dx * stress;
      let affine = stress + p_mass * C[p];

      for (let i of ti.static(ti.range(3))) {
        for (let j of ti.static(ti.range(3))) {
          let offset = [i, j];
          let dpos = (f32(offset) - fx) * dx;
          let weight = w[[i, 0]] * w[[j, 1]];
          grid_v[base + offset] +=
            weight * (p_mass * v[p] + affine.matmul(dpos));
          grid_m[base + offset] += weight * p_mass;
        }
      }
    }
    for (let I of ndrange(n_grid, n_grid)) {
      let i = I[0];
      let j = I[1];
      if (grid_m[I] > 0) {
        grid_v[I] = (1 / grid_m[I]) * grid_v[I];
        grid_v[I][1] -= dt * 50;

        if (i < 3 && grid_v[I][0] < 0) {
          //grid_v[I][0] = 0;
        }
        if (i > n_grid - 3 && grid_v[I][0] > 0) {
          //grid_v[I][0] = 0;
        }
        if (j < 3 && grid_v[I][1] < 0) {
          grid_v[I][1] = 0;
        }
        if (j > n_grid - 3 && grid_v[I][1] > 0) {
          grid_v[I][1] = 0;
        }

        //Boundary conditions

        // Inflow
        if (i < bound && grid_v[I][0] < 0) {
          grid_v[I][0] = inflow;
        }
        // Outflow
        if (i > n_grid - bound && grid_v[I][0] > 0) {
          grid_v[I][0] = inflow;
        }
        // riverbed
        let xi = i * dx;
        let rb = riverbed(xi, para_a_slider, para_b, kick_b, kick_h, kick_a);

        let y_bound = rb.y; // 0.5;
        let normal = rb.normal; // [0, 1];

        let y_j = ti.i32(y_bound * n_grid - 0.5) + 1;

        let normal_component = grid_v[I].dot(normal);

        if (j <= y_j && normal_component < 0) {
          grid_v[I] -= normal_component * normal;
        }

        // Ceiling (prevent flying too high)
        if (j > n_grid - bound && grid_v[I].y > 0) {
          grid_v[I].y = 0;
        }
      }
    }
    for (let p of range(n_particles)) {
      let base = i32(x[p] * inv_dx - 0.5);
      let fx = x[p] * inv_dx - f32(base);
      let w = [
        0.5 * (1.5 - fx) ** 2,
        0.75 - (fx - 1.0) ** 2,
        0.5 * (fx - 0.5) ** 2,
      ];
      let new_v = [0.0, 0.0];
      let new_C = [
        [0.0, 0.0],
        [0.0, 0.0],
      ];
      for (let i of ti.static(ti.range(3))) {
        for (let j of ti.static(ti.range(3))) {
          let dpos = f32([i, j]) - fx;
          let g_v = grid_v[base + [i, j]];
          let weight = w[[i, 0]] * w[[j, 1]];
          new_v = new_v + weight * g_v;
          new_C = new_C + 4 * inv_dx * weight * g_v.outerProduct(dpos);
        }
      }
      v[p] = new_v;
      C[p] = new_C;
      x[p] = x[p] + dt * new_v;

      // Respawn logic

      let rb = riverbed(x[p][0], para_a_slider, para_b, kick_b, kick_h, kick_a);
      let y_limit = rb.y;
      let normal = rb.normal;

      if (
        x[p][0] > 1.0 - 3 * dx ||
        x[p][0] < dx ||
        x[p][1] < y_limit ||
        x[p][1] > 1.0 - 3 * dx
      ) {
        x[p] = [ti.random() * 3 * dx + dx, ti.random() * river_depth];
        x[p] += [0.0, y_limit];
        v[p] = [normal.y * 2, -normal.x];
        //J[p] = 1.0;
        //C[p] = [[0.0, 0.0], [0.0, 0.0]];
      }
    }
  });

  let reset = ti.kernel(() => {
    for (let i of range(n_particles)) {
      let group_id = i32(ti.floor(i / group_size));

      x[i] = [ti.random() * 0.2 + 0.3, ti.random() * 0.2 + 0.4];
      material[i] = 0;
      //f[i] = 0 ;
      v[i] = [0, 0];
      F[i] = [
        [1, 0],
        [0, 1],
      ];
      Jp[i] = 1;
      C[i] = [
        [0, 0],
        [0, 0],
      ];
    }
  });

  let render = ti.kernel(() => {
    for (let I of ndrange(img_size, img_size)) {
      image[I] = [0.067, 0.184, 0.255, 1.0];
    }
    for (let i of range(n_particles)) {
      let pos = x[i];
      let ipos = i32(pos * img_size);
      let this_color = f32([0, 0, 0, 0]);
      if (material[i] == 0) {
        this_color = [0, 0.5, 0.5, 1.0];
      } else if (material[i] == 1) {
        this_color = [0.93, 0.33, 0.23, 1.0];
      } else if (material[i] == 2) {
        this_color = [1, 1, 1, 1.0];
      }
      image[ipos] = this_color;
    }
  });

  const getCanvasNormalizedXY = (event) => {
    var rect = htmlCanvas.getBoundingClientRect();
    if (event.touches) {
      return {
        x: (event.touches[0].clientX - rect.left) / rect.width,
        y: 1.0 - (event.touches[0].clientY - rect.top) / rect.height,
      };
    } else {
      return {
        x: (event.clientX - rect.left) / rect.width,
        y: 1.0 - (event.clientY - rect.top) / rect.height,
      };
    }
  };

  const mouseMoveListener = (event) => {
    canvasCoords = getCanvasNormalizedXY(event);
    xi = Math.floor(canvasCoords.x / n_nodes);
    if ((xi >= 0) & (xi < n_nodes)) {
      ground_y_values.set([xi], canvasCoords.y);
    }
    console.log(canvasCoords);
  };

  // document.addEventListener("mousedown", mouseDownListener);
  document.addEventListener("mousemove", mouseMoveListener);
  // document.addEventListener("mouseup", mouseupListener);

  // document.addEventListener("touchstart", mouseDownListener);
  document.addEventListener("touchmove", mouseMoveListener);
  // document.addEventListener("touchend", mouseupListener);

  const htmlCanvas = document.getElementById("result_canvas");
  htmlCanvas.width = img_size;
  htmlCanvas.height = img_size;
  const canvas = new ti.Canvas(htmlCanvas);

  reset();

  let i = 0;
  async function frame() {
    if (window.shouldStop) {
      return;
    }

    const a = parseFloat(para_a_slider_elm.value);
    //f.from_array
    for (let i = 0; i < Math.floor(2e-3 / dt); ++i) {
      substep(a, ground_y_values);
    }

    render();

    i = i + 1;
    canvas.setImage(image);
    requestAnimationFrame(frame);
  }
  await frame();
};
// This is just because StackBlitz has some weird handling of external scripts.
// Normally, you would just use `<script src="https://unpkg.com/taichi.js/dist/taichi.umd.js"></script>` in the HTML
const script = document.createElement("script");
script.addEventListener("load", function () {
  main();
});

script.src = "https://unpkg.com/taichi.js/dist/taichi.umd.js";
// Append to the `head` element
document.head.appendChild(script);
